File: Tugas-Kelompok-3/soal1.cpp
/*	
Nama	     : Rifqi Muhadzib Ahdan - M. Ikhsan Fadhilah - M. Zidan Yohanza - M. Faris Ary Prasetyo
NPM		     : 140810210029 - 140810210003 - 140810210021 -140810190056
Kelas		 : A
Tanggal	     : 19 Oktober 2022
Deskripsi    : Menentukan huruf mutu dan lama durasi ujian masing-masing mahasiswa yang dikumpukan dengan array of class dan mencari nilai tertinggi, terenda, dan rata-rata nilai semua mahasiswa
*/

#include <iostream>
#include <iomanip>

class Waktu{
    private:
        int jam, menit, detik;

    public:
        Waktu(){
            this->jam = 0;
            this->menit = 0;
            this->detik = 0;
        }

        Waktu(int jam, int menit, int detik){
            this->jam = jam;
            this->menit = menit;
            this->detik = detik;
        }

        // Setters
        void setJam(int jam){
            this->jam = jam;
        }
        void setMenit(int menit){
            this->menit = menit;
        }
        void setDetik(int detik){
            this->detik = detik;
        }
        void setWaktu(int jam, int menit, int detik){
            this->jam = jam;
            this->menit = menit;
            this->detik = detik;
        }

        // Getters
        int getJam(){
            return this->jam;
        }
        int getMenit(){
            return this->menit;
        }
        int getDetik(){
            return this->detik;
        }

        Waktu getDurasi(Waktu awal, Waktu akhir){
            int detikAwal = awal.getJam() * 3600 + awal.getMenit() * 60 + awal.getDetik();
            int detikAkhir = akhir.getJam() * 3600 + akhir.getMenit() * 60 + akhir.getDetik();
            int durasi = detikAkhir - detikAwal;

            int jam = durasi / 3600;
            int menit = (durasi % 3600) / 60;
            int detik = (durasi % 3600) % 60;

            return Waktu(jam, menit, detik);
        }

        std::string printWaktu(){
            return std::to_string(this->jam) + ":" + std::to_string(this->menit) + ":" + std::to_string(this->detik);
        }

        bool fixTime(Waktu waktu){
            bool durasiErr;
            int detikMulai = this->jam * 3600 + this->menit * 60 + this->detik;
            int detikSelesai = waktu.getJam() * 3600 + waktu.getMenit() * 60 + waktu.getDetik();

            if (detikMulai > detikSelesai){
                durasiErr = true;
                std::cout << "Waktu selesai harus lebih besar dari waktu mulai" << std::endl;
            } else {
                durasiErr = false;
            }
            return durasiErr;
        }

        Waktu inputWaktu(){
            bool waktuValidation = false;
            int jam, menit, detik;

            do{
                std::cout << "Jam : ";
                std::cin >> jam;

                if(jam < 0 || jam > 23){
                    std::cout << "Waktu tidak valid" << std::endl;
                    continue;
                }

                std::cout << "Menit : ";
                std::cin >> menit;

                if(menit < 0 || menit > 59){
                    std::cout << "Waktu tidak valid" << std::endl;
                    continue;
                }

                std::cout << "Detik : ";
                std::cin >> detik;

                if(detik < 0 || detik > 59){
                    std::cout << "Waktu tidak valid" << std::endl;
                    continue;
                }

                waktuValidation = true;

            } while (!waktuValidation);

            return Waktu(jam, menit, detik);
        }
};

class Mahasiswa
{
private:
    std::string nama, npm;
    Waktu awal, akhir;
    int nilai;

public:
    Mahasiswa(){
        this->nama = "";
        this->npm = "";
        this->nilai = 0;
        this->awal = Waktu(0, 0, 0);
        this->akhir = Waktu(0, 0, 0);
    }

    Mahasiswa(std::string nama, std::string npm, Waktu awal, Waktu akhir, int nilai){
        this->nama = nama;
        this->npm = npm;
        this->awal = awal;
        this->akhir = akhir;
        this->nilai = nilai;
    }

    void setNama(std::string nama){
        this->nama = nama;
    }

    void setNpm(std::string npm){
        this->npm = npm;
    }

    void setAwal(Waktu awal){
        this->awal = awal;
    }

    void setAkhir(Waktu akhir){
        this->akhir = akhir;
    }

    void setNilai(int nilai){
        this->nilai = nilai;
    }

    std::string getNama(){
        return this->nama;
    }

    std::string getNpm(){
        return this->npm;
    }

    Waktu getAwal(){
        return this->awal;
    }

    Waktu getAkhir(){
        return this->akhir;
    }

    int getNilai(){
        return this->nilai;
    }

    Waktu getWaktuDurasi(){
        return this->awal.getDurasi(this->awal, this->akhir);
    }

    char getHurufMutu() {
        if (nilai >= 80 && nilai <= 100){
            return 'A';
        } else if (nilai >= 68 && nilai < 80){
            return 'B';
        } else if (nilai >= 55 && nilai < 68){
            return 'C';
        } else if (nilai >= 45 && nilai < 55){
            return 'D';
        } else if (nilai >= 0 && nilai < 45){
            return 'E';
        } else {
            return ' ';
        }
    }

    int getHighScore(Mahasiswa mhs[], int n){
        int max = 0;
        for (int i = 0; i < n; i++){
            if (mhs[i].getNilai() > max){
                max = mhs[i].getNilai();
            }
        }
        return max;
    }

    int getLowScore(Mahasiswa mhs[], int n){
        int min = 100;
        for (int i = 0; i < n; i++){
            if (mhs[i].getNilai() < min){
                min = mhs[i].getNilai();
            }
        }
        return min;
    }

    double getMeanScore(Mahasiswa mhs[], int n){
        int total = 0;
        for (int i = 0; i < n; i++){
            total += mhs[i].getNilai();
        }
        return total / n;
    }

    std::string kelulusan(){
        if (getNilai() >= 55){
            return "Lulus";
        } else {
            return "Gagal";
        }
    }

    void inputMahasiswa(Mahasiswa mhs[], int n){
        std::string nama, npm;
        Waktu awal = Waktu(0, 0, 0);
        Waktu akhir = Waktu(0, 0, 0);
        int nilai;

        for (int i = 0; i < n; i++){
            std::cout << "\nMahasiswa ke-" << i + 1 << std::endl;
            std::cout << "Nama : ";
            std::cin.ignore();
            std::getline(std::cin, nama);
            std::cout << "NPM : ";
            std::cin >> npm;

            do{
                std::cout << "===Waktu Mulai===" << std::endl;
                awal = awal.inputWaktu();

                std::cout << "===Waktu Selesai===" << std::endl;
                akhir = akhir.inputWaktu();

            } while (awal.fixTime(akhir));
            std::cout << "Nilai : ";
            std::cin >> nilai;

            mhs[i] = Mahasiswa(nama, npm, awal, akhir, nilai);
        }
    }

    void printData(Mahasiswa mhs[], int n)
    {   
        std::cout << "\n============================================= DATA MAHASISWA =================================================================" << std::endl;
        std::cout << std::setw(5) << "No" << std::setw(20) << "Nama" << std::setw(20) << "NPM" << std::setw(15) << "Nilai Ujian" << std::setw(10) << "HM" << std::setw(15) << "Waktu Mulai" << std::setw(15) << "Waktu Selesai" << std::setw(15) << "Waktu Durasi" << std::setw(10) << "Status" << std::endl;
        for (int i = 0; i < n; i++)
        {
            std::cout << std::setw(5) << i + 1 << std::setw(20) << mhs[i].getNama() << std::setw(20) << mhs[i].getNpm() << std::setw(15) << mhs[i].getNilai() << std::setw(10) << mhs[i].getHurufMutu() << std::setw(15) << mhs[i].getAwal().printWaktu() << std::setw(15) << mhs[i].getAkhir().printWaktu() << std::setw(15) << mhs[i].getWaktuDurasi().printWaktu() << std::setw(10) << mhs[i].kelulusan() << std::endl;
        }
        std::cout << "==============================================================================================================================" << std::endl;
        std::cout << "Rata-rata Nilai : " << mhs[0].getMeanScore(mhs, n) << std::endl;
        std::cout << "Nilai Tertinggi: " << mhs[0].getHighScore(mhs, n) << std::endl;
        std::cout << "Nilai Terendah: " << mhs[0].getLowScore(mhs, n) << std::endl;

    }
};

main()
{
    int n;
    std::cout << "Masukkan jumlah mahasiswa : ";
    std::cin >> n;

    Mahasiswa mhs[n];
    mhs->inputMahasiswa(mhs, n);
    mhs->printData(mhs, n);
}

File: Tugas-Kelompok-3/soal1.py
#  Nama	         : Rifqi Muhadzib Ahdan - M. Ikhsan Fadhilah - M. Zidan Yohanza - M. Faris Ary Prasetyo
#  NPM		     : 140810210029 - 140810210003 - 140810210021 -140810190056
#  Kelas		 : A
#  Tanggal	     : 19 Oktober 2022
#  Deskripsi     : Menentukan huruf mutu dan lama durasi ujian masing-masing mahasiswa yang dikumpukan dengan array of object
#                  Serta mencari nilai tertinggi, terendah, dan rata-rata nilai semua mahasiswa

class daftarMahasiswa :
    def __init__(self, jumlah=0):
        self.__jumlah = jumlah
        self.__arrMahasiswa = []
     # Setters
    def setJumlah(self, jumlah):
        self.__jumlah = jumlah
    def isiArr(self):
        for i in range(0, self.__jumlah) :
            tempMahasiswa = Mahasiswa()
            print("Mahasiswa ke-" + str((i)+1))
            tempMahasiswa.inputMahasiswa()
            self.__arrMahasiswa.append(tempMahasiswa)
    def printDaftar(self):
        totalNilai = 0
        max = 0
        min = 100

        print("===OUTPUT===")
        print("\t\t\t\t\tDaftar Nilai Ujian dan Lama Ujian\n\n")
        print("{:<5s} {:<13s} {:<15s} {:<20s} {:<10s} {:<15s} {:<10s} {:<10s} {:<10s}".format("NO", "NAMA", "NPM", "NILAI UJIAN", "HM", "STATUS", "MULAI", "SELESAI", "LAMA"))
        for x in range(0, 120):
            print("=", end='')
        print()
        for i in range(0, self.__jumlah):
            # Proses cari max, min, dan rerata
            temp = self.__arrMahasiswa[i]
            if(temp.getNilai() > max): max = temp.getNilai()
            if(temp.getNilai() < min): min = temp.getNilai()
            totalNilai += temp.getNilai()

            # Row per mahasiswa
            temp.toTableRow((i+1))
        print("\nNilai Tertinggi : ", max)
        print("Nilai Terendah  : ", min)
        print("Rata-rata nilai : ", (totalNilai/self.__jumlah))

class Mahasiswa :
    # Constructors
    def __init__(self, nama=None, npm=None, nilai=0) :
        self.__nama = nama
        self.__npm = npm
        self.__nilai = nilai
        self.__awal = Waktu()
        self.__akhir = Waktu()

    # Setters
    def setNama(self, nama) :
        self.__nama = nama
    def setNpm(self, npm) :
        self.__npm = npm
    def setNilai(self, nilai) :
        self.__nilai = nilai
    def setAwal(self, awal) :
        self.__awal = awal
    def setAkhir(self, akhir) :
        self.__akhir = akhir

    # Getters
    def getNama(self) :
        return self.__nama
    def getNpm(self) :
        return self.__npm
    def getNilai(self) :
        return self.__nilai
    def getAwal(self) :
        return self.__awal
    def getAkhir(self) :
        return self.__akhir
    
    # fungsional
    def inputMahasiswa(self):
        self.setNama(input("Nama          : "))
        self.setNpm(input("NPM           : "))
        self.setNilai(int(input("Nilai         : ")))
        self.getAwal().setWaktuDariMilitary(input("Waktu Mulai   : "))
        self.getAwal().fixTime()
        self.getAkhir().setWaktuDariMilitary(input("Waktu Selesai : "))
        self.getAkhir().fixTime()
    def hurufMutu(self) :
        if (self.__nilai >= 0 and self.__nilai < 45) :
            return 'E'
        elif (self.__nilai >= 45 and self.__nilai < 55) :
            return 'D'
        elif (self.__nilai >= 55 and self.__nilai < 68) :
            return 'C'
        elif (self.__nilai >= 68 and self.__nilai < 80) :
            return 'B'
        else :
            return 'A'
    def kelulusan(self) :
        if(self.__nilai >= 55): return "Lulus"
        else : return "Tidak Lulus"
    def toTableRow(self, urutan) :
        waktuDurasi = Waktu()
        # Passing object cara 1 (without return value)
        waktuDurasi.hitungDurasi(self.__awal, self.__akhir)
        waktuDurasi.fixTime()

        print("{:<5d} {:<13s} {:<15s} {:<20d} {:<10s} {:<15s} {:<10s} {:<10s} {:<10s}"
                .format(urutan, self.__nama, self.__npm, self.__nilai, self.hurufMutu(), 
                self.kelulusan(), self.__awal.printWaktu(), self.__akhir.printWaktu(), 
                waktuDurasi.printWaktu()))

class Waktu:
    # Constructor
    def __init__(self, jam=0, menit=0, detik=0):
        self.__jam = jam
        self.__menit = menit
        self.__detik = detik
    # Setters
    def setJam(self, jam):
        self.__jam = jam
    def setMenit(self, menit):
        self.__menit = menit
    def setDetik(self, detik):
        self.__detik = detik
    def setWaktuDariMilitary(self, waktu):
        arrWaktu = waktu.split(":")
        self.__jam = int(arrWaktu[0])
        self.__menit = int(arrWaktu[1])
        self.__detik = int(arrWaktu[2])
    # Getters
    def getJam(self):
        return self.__jam
    def getMenit(self):
        return self.__menit
    def getDetik(self):
        return self.__detik
    # Fungsi
    def fixTime(self) :
        if (self.__detik >= 60) :
            self.__menit += (self.__detik / 60)
            self.__detik %= 60
        
        if (self.__menit >= 60) :
            self.__jam += (self.__menit / 60)
            self.__menit %= 60
        
        if (self.__jam >= 24) :
            self.__jam %= 24
    # Passing object cara 1 
    def hitungDurasi(self, awal, akhir) :
        self.__detik += (akhir.getDetik() - awal.getDetik())
        if (self.__detik < 0) :
            self.__menit -= 1
            self.__detik += 60

        self.__menit += (akhir.getMenit() - awal.getMenit())
        if (self.__menit < 0) :
            self.__jam -= 1
            self.__menit += 60

        self.__jam += (akhir.getJam() - awal.getJam())
        if (self.__jam < 0) :
            self.__jam += 24 
    def printWaktu(self) :
        return format("%02d:%02d:%02d" %(self.__jam, self.__menit, self.__detik))
    
# main
mhs = daftarMahasiswa()
mhs.setJumlah(int(input("Input jumlah mahasiswa : ")))
mhs.isiArr()
mhs.printDaftar()